core-rust/state-manager/src/protocol/protocol_state.rs

```diff
@@ -0,0 +1,511 @@
+use radix_engine::blueprints::consensus_manager::EpochChangeEvent;
+use radix_engine::prelude::{ScryptoCategorize, ScryptoDecode, ScryptoEncode};
+use std::cmp::Ordering;
+use std::collections::BTreeMap;
+
+use radix_engine_common::constants::CONSENSUS_MANAGER;
+
+use radix_engine_common::math::Decimal;
+use radix_engine_common::prelude::scrypto_decode;
+
+use radix_engine_common::types::Epoch;
+use radix_engine_interface::api::ModuleId;
+use radix_engine_interface::prelude::CheckedMul;
+use radix_engine_interface::prelude::Emitter;
+use tracing::info;
+
+use crate::traits::{IterableProofStore, QueryableProofStore, QueryableTransactionStore};
+use crate::ProtocolUpdateEnactmentCondition::{
+    EnactUnconditionallyAtEpoch, EnactWhenSupportedAndWithinBounds,
+};
+use crate::{
+    LocalTransactionReceipt, ProtocolConfig, ProtocolUpdate, SignalledReadinessThreshold,
+    StateVersion,
+};
+
+// This file contains types and utilities for
+// managing the (dynamic) protocol state of a running node.
+
+#[derive(Clone, Debug, Eq, PartialEq, ScryptoCategorize, ScryptoEncode, ScryptoDecode)]
+pub struct ProtocolState {
+    pub current_epoch: Option<Epoch>,
+    pub current_protocol_version: String,
+    pub enacted_protocol_updates: BTreeMap<StateVersion, String>,
+    pub unenacted_protocol_updates: Vec<UnenactedProtocolUpdate>,
```

Comment:
Yeah, I think it's actually a better name (I considered it, but rejected for the reason I no longer remember).

Reply:
I know, I considered it in "ProtocolConfig", but "pending" doesn't really fit there because at the (static) configuration level the node doesn't know if it's "pending" or "past". Okay to use it in (dynamic) state though

Reply:
Cool, I like `applied_protocol_updates` and `pending_protocol_updates`

Reply:
(renamed)